### PR TITLE
OSDOCS-8615: Configuring PID limits

### PR DESCRIPTION
**Pull request by bhardesty**
This PR documents how to configure PID limits in ROSA, using the ROSA CLI. When the UI is added, this content will be expanded to include a separate procedure for setting PID limits in OCM.

Version(s):
enterprise-4.14+

Issue:
https://issues.redhat.com/browse/OSDOCS-8615

Link to docs preview:
https://69352--ocpdocs-pr.netlify.app/openshift-rosa/latest/rosa_cluster_admin/rosa-configuring-pid-limits

QE review:
- [ ] QE has approved this change.

### CODE REVIEW

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
++
+--
+** If this is the first time you are changing the default PID limit, create the kubeletconfig and set the `--pod-pids-limit` value. 
++
+[source,terminal]
+----
+$ rosa create kubeletconfig -c <cluster_name> --pod-pids-limit=<value>
+----
++
+For example, to set a maximum of 16,384 PIDs per pod for cluster my-cluster when you have never changed the default value:
```
Comment (xueli181114) [inline modules/setting-higher-pid-limit-on-existing-cluster.adoc:33]:
```suggestion
For example, to set a maximum of 16,384 PIDs per pod for cluster my-cluster when you have never created a kubeletconfig
```

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
++
+[source,terminal]
+----
+$ rosa create kubeletconfig -c my-cluster --pod-pids-limit=16384
+----
+
+** If you previously changed the default PID limit, edit the existing kubeletconfig and set the `--pod-pids-limit` value.
```
Comment (xueli181114) [inline modules/setting-higher-pid-limit-on-existing-cluster.adoc:40]:
```suggestion
** If you previously created a kubeletconfig, edit the existing kubeletconfig and set the `--pod-pids-limit` value.
```

File: modules/understanding-process-id-limits.adoc [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+:_mod-docs-content-type: CONCEPT
+[id="understanding-process-id-limits_{context}"]
+= Understanding process ID limits
+
+A process identifier (PID) is a unique identifier assigned by the Linux kernel to each process or thread currently running on a system. The number of processes that can run simultaneously on a system is limited to 4,194,304 by the Linux kernel, and may additionally be affected by limited access to other system resources such as memory, CPU, and disk space.
+
+In {product-title}, there are two supported limits for PID usage to consider when you schedule work on your cluster:
+
+Maximum number of PIDs per pod (default: 4,096 in {product-title} 4.11 and higher)::
```
Comment (shipsing) [inline modules/understanding-process-id-limits.adoc:13]:
```suggestion
Maximum number of PIDs per pod (default: 4,096 in {product-title} 4.11 and later)::
```

File: modules/understanding-process-id-limits.adoc [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+This is controlled by the `podPidsLimit` parameter set on the node.
+
+Maximum number of PIDs per node (default depends on link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html-single/nodes/index#nodes-nodes-resources-configuring[node resources])::
+In OpenShift, this is controlled by the link:https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#system-reserved[`--system-reserved`] parameter, which reserves PIDs on each node based on the total resources of the node.
```
Comment (shipsing) [inline modules/understanding-process-id-limits.adoc:17]: @bburt-rh, Is the formatting okay? I think the writer wants to highlight these points using a description list.

File: rosa_cluster_admin/rosa-configuring-pid-limits.adoc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+toc::[]
+
+include::modules/understanding-process-id-limits.adoc[leveloffset=+1]
+
+include::modules/risks-setting-higher-process-id-limits.adoc[leveloffset=+1]
+
+[role="_additional-resources"]
+[id="additional-resources_pid-limit-risks"]
+.Additional resources
```
[the remark this answers is not in the document]
Comment (shipsing) [reply rosa_cluster_admin/rosa-configuring-pid-limits.adoc:26]: @bburt-rh A query: If we use the following formatting to indicate that the resource applies specifically to the module and not to the assembly:
[role="_additional-resources"]
.Additional resources

Do we need to include this formatting in that specific module file or in the associated assembly file?

File: modules/understanding-process-id-limits.adoc [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+When a pod exceeds the allowed maximum number of PIDs per pod, the pod may stop functioning correctly and may be evicted from the node. See the Kubernetes documentation for more information on eviction: link:https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals-and-thresholds[Eviction signals and thresholds].
+
+When a node exceeds the allowed maximum number of PIDs per node, the node can become unstable because new processes cannot have PIDs assigned. If existing processes cannot complete without creating additional processes, the entire node may become unusable and require reboot. This may result in data loss, depending on the processes and applications being run. Customer administrators and Red Hat Site Reliability Engineering are notified when this threshold is reached, and you will notice a "Worker node is experiencing PIDPressure" warning in cluster logs.
```
Comment (shipsing) [inline modules/understanding-process-id-limits.adoc:21]:
```suggestion
When a node exceeds the allowed maximum number of PIDs per node, the node can become unstable because new processes cannot have PIDs assigned. If existing processes cannot complete without creating additional processes, the entire node may become unusable and require reboot. This may result in data loss, depending on the processes and applications being run. Customer administrators and Red Hat Site Reliability Engineering are notified when this threshold is reached, and you will notice a `Worker node is experiencing PIDPressure` warning in cluster logs.
```
We need to enclose a warning in backticks: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#system-messages

**Review by shipsing — CHANGES_REQUESTED** [review]

**Comment by shipsing** [comment]
@bhardesty, Nice work. I have provided a few suggestions for you to consider. Please have a look at them.

Since i am new to the peer review squad, i would ask @bburt-rh to give my review the go-ahead. Please wait until Brain applies the peer-review-done label.

File: modules/risks-setting-higher-process-id-limits.adoc [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+[id="risks-setting-higher-process-id-limits_{context}"]
+= Risks of setting higher process ID limits for {product-title} pods
+
+The `podPidsLimit` parameter for a pod controls the maximum number of processes and threads that can run simultaneously in that pod.
+
+You can increase the value for `podPidsLimit` from the default of 4,096 to a maximum of 16,384. Changing this value may incur downtime for applications, because changing the `podPidsLimit` requires rebooting the affected node.
+
+If you are running a large number of pods per node, and you have a high `podPidsLimit` value on your nodes, you risk exceeding the PID maximum for the node.
+
+To find the maximum number of pods that you can run simultaneously on a single node without exceeding the PID maximum for the node, divide 3,650,000 by your `podPidsLimit` value. For example, if your `podPidsLimit` value is 1,048,576, and you expect the pods to use close to that number of process IDs, you can safely run three pods on a single node.
```
Comment (bchandra-ocp) [inline modules/risks-setting-higher-process-id-limits.adoc:15]: I love the idea of giving an example. Can we adjust the example to align with maximum supported value of 16,384? Introducing a number 1,048,576 may confuse or wrongly give an impression that podPidsLimit support this value; we've just above mentioned that the maximum generally allowed is 16,384. Perhaps we use a number within the allowed range?

File: modules/risks-setting-higher-process-id-limits.adoc [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+[NOTE]
+====
+Memory, CPU, and available storage can also limit the maximum number of pods that can run simultaneously, even when `podPidsLimit` is set appropriately. For more information, see "Planning your environment" and "Limits and scalability".
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply modules/risks-setting-higher-process-id-limits.adoc:19]: @shipsing Our guidance on this could be clarified for sure, but the ISG suggests that we use double-quotation marks: https://www.ibm.com/docs/en/ibm-style?topic=format-highlighting. The table is kind of hard to browse, but it mentions that "Topic titles in IBM Documentation" should be highlighted in double-quotation marks.

So, absent any other guidance, I think the quotation marks can be left as-is.

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="setting-higher-pid-limit-on-existing-cluster_{context}"]
+= Setting a higher PID limit on an existing {product-title} cluster
+
+You can set a higher `podPidsLimit` on an existing {product-title} cluster by creating or editing a kubeletconfig that changes the `--pod-pids-limit` parameter.
+
+[IMPORTANT]
+====
+Changing the `podPidsLimit` on an existing cluster will trigger non-control plane nodes in the cluster to reboot one at a time. Red Hat recommends making this change outside of peak usage hours for your cluster, and avoiding upgrading or hibernating your cluster until all nodes have rebooted.
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply modules/setting-higher-pid-limit-on-existing-cluster.adoc:13]:
```suggestion
Changing the `podPidsLimit` on an existing cluster will trigger non-control plane nodes in the cluster to reboot one at a time. Make this change outside of peak usage hours for your cluster and avoid upgrading or hibernating your cluster until all nodes have rebooted.
```
Per the [SSG](https://redhat-documentation.github.io/supplementary-style-guide/#recommend), avoid "recommends" completely.

'Instead of "Red Hat recommends", direct users to take the recommended action. This allows Red Hat to be more prescriptive in documentation and prevent any user uncertainty, and is easier for upstream or downstream coordinated efforts.'

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc (hunk 1 above)
Comment (bburt-rh) [reply modules/setting-higher-pid-limit-on-existing-cluster.adoc:33]:
```suggestion
For example, to set a maximum of 16,384 PIDs per pod for a cluster named `my-cluster` when you have never changed the default value, run the following command:
```
Revise this introductory phrase to be a complete sentence and to introduce the command using the [OCP docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command).

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc (hunk 2 above)
Comment (bburt-rh) [inline modules/setting-higher-pid-limit-on-existing-cluster.adoc:40]:
```suggestion
** If you previously changed the default PID limit, edit the existing `kubeletconfig` object and set the `--pod-pids-limit` value.
```
- `kubeletconfig` is a literal and needs a descriptive noun. I'm no sure if it's an object, a file, or a config map or what, though.

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+----
+--
++
+A cluster-wide rolling reboot of worker nodes is triggered.
+
+.Verification
+
+When each node in the cluster has rebooted, you can verify that the new setting is in place.
+
+* Check the Pod Pids limit in the kubeletconfig:
```
Comment (bburt-rh) [inline modules/setting-higher-pid-limit-on-existing-cluster.adoc:54]:
```suggestion
* Check the Pod Pids limit in the `kubeletconfig` [object???], by running the following command:
```
- Format `kubeletconfig` as a literal.
- Revise this introductory phrase to introduce the command using the [OCP docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command).

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
++
+[source,terminal]
+----
+$ rosa describe kubeletconfig --cluster=<cluster_name>
+----
++
+You should see the new Pod Pids limit in the output. For example:
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply modules/setting-higher-pid-limit-on-existing-cluster.adoc:61]: Yes. @shipsing is correct. This line can be completely omitted. 

- For future reference, we need a complete sentence as an introduction. `For example:` by itself does not suffice.

```suggestion
The new Pod Pids limit appears in the output, as shown in the following example:
```
- Also for future reference, [per the ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#s), avoid the word "should" and try to rewrite in a more direct way; "should" implies some doubt as to whether or not the action will have the desired result.

File: modules/understanding-process-id-limits.adoc [hunk 12]
```diff
@@ -0,0 +1,21 @@
+// Module included in the following assemblies:
+//
+// * rosa_cluster_admin/rosa-configuring-pid-limits.adoc
+
+:_mod-docs-content-type: CONCEPT
+[id="understanding-process-id-limits_{context}"]
+= Understanding process ID limits
+
+A process identifier (PID) is a unique identifier assigned by the Linux kernel to each process or thread currently running on a system. The number of processes that can run simultaneously on a system is limited to 4,194,304 by the Linux kernel, and may additionally be affected by limited access to other system resources such as memory, CPU, and disk space.
```
Comment (bburt-rh) [inline modules/understanding-process-id-limits.adoc:9]:
```suggestion
A process identifier (PID) is a unique identifier assigned by the Linux kernel to each process or thread currently running on a system. The number of processes that can run simultaneously on a system is limited to 4,194,304 by the Linux kernel. This number might also be affected by limited access to other system resources such as memory, CPU, and disk space.
```
- [Per the ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#m), only use "may" to indicate permission.
- For minimalism and clarity, break the long 2nd sentence into two sentences. 

cc @shipsing

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 13]
```diff
@@ -0,0 +1,67 @@
+// Module included in the following assemblies:
+//
+// * rosa_cluster_admin/rosa-configuring-pid-limits.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="setting-higher-pid-limit-on-existing-cluster_{context}"]
+= Setting a higher PID limit on an existing {product-title} cluster
+
+You can set a higher `podPidsLimit` on an existing {product-title} cluster by creating or editing a kubeletconfig that changes the `--pod-pids-limit` parameter.
```
[the remark this answers is not in the document]
Comment (bchandra-ocp) [reply modules/setting-higher-pid-limit-on-existing-cluster.adoc:9]: I'd also suggest to include a guidance to use status of machine config pool to ensure that the reboot is completed? The instruction could either be linked to https://docs.openshift.com/container-platform/4.8/post_installation_configuration/machine-configuration-tasks.html#checking-mco-status_post-install-machine-configuration-tasks or restated for better UX.

**Review by bchandra-ocp — CHANGES_REQUESTED** [review]
Can we also add an instruction to have customer watch for status of machine config pool https://docs.openshift.com/container-platform/4.8/post_installation_configuration/machine-configuration-tasks.html#checking-mco-status_post-install-machine-configuration-tasks to ensure nodes are rebooted. 
The ROSA command from OCM doesn't provide status yet. 
Even linking to OpenShift documentation is fine.

File: modules/understanding-process-id-limits.adoc [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+In {product-title}, there are two supported limits for PID usage to consider when you schedule work on your cluster:
```
Comment (bburt-rh) [inline modules/understanding-process-id-limits.adoc:11]:
```suggestion
In {product-title}, consider these two supported limits for PID usage before you schedule work on your cluster:
```
Per [the ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#t), avoid "there are" at the beginning of an independent clause.

cc @shipsing

File: modules/understanding-process-id-limits.adoc [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+Maximum number of PIDs per pod (default: 4,096 in {product-title} 4.11 and higher)::
+This is controlled by the `podPidsLimit` parameter set on the node.
```
Comment (bburt-rh) [inline modules/understanding-process-id-limits.adoc:14]:
```suggestion
This limit is controlled by the `podPidsLimit` parameter set on the node.
```
- Per the ISG, avoid ambigous pronoun references. It's unclear here what "this" refers to. This what?

cc @shipsing

File: modules/understanding-process-id-limits.adoc (hunk 4 above)
Comment (bburt-rh) [reply modules/understanding-process-id-limits.adoc:17]: @shipsing @bhardesty - The two items here should be formatted as a bulleted list.

File: modules/understanding-process-id-limits.adoc (hunk 3 above)
Comment (bburt-rh) [reply modules/understanding-process-id-limits.adoc:13]:
```suggestion
* Maximum number of PIDs per pod. The default value is 4,096 in {product-title} 4.11 and later)
```
A few other changes in addition to the change by @shipsing that follows [the ISG guidance about referring to product versions](https://www.ibm.com/docs/en/ibm-style?topic=references-product-versions#references-to-earlier-or-later-versions).

File: modules/understanding-process-id-limits.adoc (hunk 4 above)
Comment (bburt-rh) [reply modules/understanding-process-id-limits.adoc:17]:
```suggestion
Maximum number of PIDs per pod (default: 4,096 in {product-title} 4.11 and higher)::
This is controlled by the `podPidsLimit` parameter set on the node.

* Maximum number of PIDs per node. The default value depends on link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html-single/nodes/index#nodes-nodes-resources-configuring[node resources]).
In {product-title}, this value is controlled by the link:https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#system-reserved[`--system-reserved`] parameter, which reserves PIDs on each node based on the total resources of the node.
```
- Format as a bullet point.
- Replace "OpenShift" with the correct attribute.
- Rewrite to avoid ambiguous pronoun reference.
- Also, as a point for future reference: Generally in the non-ROSA OCP docs, we don't cross-reference other OCP docs by using direct links the published HTML; we should always use xrefs to adoc files, though these xrefs can't be used in module files. However, for the ROSA docs, there's an exception because we can't link between different distros, such as from ROSA to OpenShift Container Platform. Thus, these can be external links and not cross-references. These external links use `link` instead of `xref` and need to be in distro-specific files or conditionalized to apply to the relevant distros.

File: modules/understanding-process-id-limits.adoc [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+Maximum number of PIDs per node (default depends on link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html-single/nodes/index#nodes-nodes-resources-configuring[node resources])::
+In OpenShift, this is controlled by the link:https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#system-reserved[`--system-reserved`] parameter, which reserves PIDs on each node based on the total resources of the node.
+
+When a pod exceeds the allowed maximum number of PIDs per pod, the pod may stop functioning correctly and may be evicted from the node. See the Kubernetes documentation for more information on eviction: link:https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals-and-thresholds[Eviction signals and thresholds].
```
Comment (bburt-rh) [inline modules/understanding-process-id-limits.adoc:19]:
```suggestion
When a pod exceeds the allowed maximum number of PIDs per pod, the pod might stop functioning correctly and might be evicted from the node. See link:https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals-and-thresholds[the Kubernetes documentation for eviction signals and thresholds] for more information.
```
- Don't use "may" to mean "might". Use "may" only to indicate permission
- Revise the "for more information" sentence for minimalism. My suggestion is one possibility.

File: modules/understanding-process-id-limits.adoc (hunk 6 above)
Comment (bburt-rh) [inline modules/understanding-process-id-limits.adoc:21]:
```suggestion
When a node exceeds the allowed maximum number of PIDs per node, the node can become unstable because new processes cannot have PIDs assigned. If existing processes cannot complete without creating additional processes, the entire node can become unusable and require reboot. This situation can result in data loss, depending on the processes and applications being run. Customer administrators and Red Hat Site Reliability Engineering are notified when this threshold is reached, and a `Worker node is experiencing PIDPressure` warning will appear in the cluster logs.
```
- Don't use "may" except to indicate permission.
- Don't use "this" without a clear reference.

File: rosa_cluster_admin/rosa-configuring-pid-limits.adoc [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+// something similar.
+//
+// Also, this content theoretically applies to OSD as well, but the PID limits
+// can only be configured through the ROSA CLI. When other methods are added for
+// configuring PID limits, this assembly and its modules can be evaluated for 
+// inclusion in the OSD docs.
+
+:_mod-docs-content-type: ASSEMBLY
+[id="rosa-configuring-pid-limits"]
+= Configuring PID limits
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply rosa_cluster_admin/rosa-configuring-pid-limits.adoc:14]: @shipsing is correct. You need some intro test here.

File: rosa_cluster_admin/rosa-configuring-pid-limits.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bburt-rh) [reply rosa_cluster_admin/rosa-configuring-pid-limits.adoc:26]: Per the link that @shipsing cites, in assemblies, an "Additional resources" section should be formatted as follows:

`
[role="_additional-resources"]
[id="additional-resources_pid-limit-risks"]
== Additional resources
`

and not 

`.Additional resources`

As for where to put the "Additional resources" section (in the assembly or the module?), it all depends on where you want the additional resources section to appear. It's at the writer's discretion AFAIU.

File: snippets/pid-limits.adoc [hunk 18]
```diff
@@ -15,9 +15,17 @@
 
 :_mod-docs-content-type: SNIPPET
 
+// The PID limits are only configurable through the ROSA CLI. Once they are configurable through OCM, the conditionalization can be removed.
+
 [IMPORTANT]
 ====
-As of the {product-title} versions 4.8.35, 4.9.26, 4.10.6, the {product-title} default per-pod pid limit is `4096`. If you want to enable this PID limit, you must upgrade your {product-title} clusters to these versions or later. {product-title} clusters with prior versions use a default PID limit of `1024`.
+As of {product-title} versions 4.11, the default per-pod PID limit is `4096`. If you want to enable this PID limit, you must upgrade your {product-title} clusters to these versions or later. {product-title} clusters with prior versions use a default PID limit of `1024`.
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply snippets/pid-limits.adoc:22]: Also, it should be `earlier versions` not `prior versions`. 

I would also say `{product-title} clusters running on earlier versions`

File: _topic_maps/_topic_map_rosa.yml [hunk 19]
```diff
@@ -528,6 +528,8 @@ Topics:
     File: rosa-nodes-about-autoscaling-nodes
   - Name: Configuring cluster memory to meet container memory and risk requirements
     File: nodes-cluster-resource-configure
+- Name: Configuring PID limits
+  File: rosa-configuring-pid-limits
```
[the remark this answers is not in the document]
Comment (bburt-rh) [reply _topic_maps/_topic_map_rosa.yml:532]: It depends on where the writer wants the topic to appear in the TOC. The "Configuring PID limits" is now at this level:

image

You would only indent the new entry further if you wanted the topic to be below

File: modules/risks-setting-higher-process-id-limits.adoc (hunk 8 above)
Comment (bburt-rh) [inline modules/risks-setting-higher-process-id-limits.adoc:19]:
```suggestion
Memory, CPU, and available storage also limits the maximum number of pods that can run simultaneously, even when the `podPidsLimit` value is set appropriately. For more information, see "Planning your environment" and "Limits and scalability".
```

File: modules/setting-higher-pid-limit-on-existing-cluster.adoc [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+* You have logged in to your Red Hat account by using the ROSA CLI.
+* You have installed the OpenShift CLI (`oc`).
+
+.Procedure
+
+* Create or edit the kubeletconfig to change the PID limit.
++
+--
+** If this is the first time you are changing the default PID limit, create the kubeletconfig and set the `--pod-pids-limit` value.
```
Comment (bburt-rh) [inline modules/setting-higher-pid-limit-on-existing-cluster.adoc:26]:
```suggestion
** If this is the first time you are changing the default PID limit, create the kubeletconfig and set the `--pod-pids-limit` value by running the following command: 
```
- Follow OCP guidelines for introducing a terminal command: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command

**Comment by bburt-rh** [comment]
@bhardesty - Thanks for the PR. Please revise this PR based on the review comments and suggestions, and ping us (me and @shipsing) again so that we can take another look and give it the peer-review-done tag. Thanks again!

File: _topic_maps/_topic_map_rosa.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (bhardesty) [reply _topic_maps/_topic_map_rosa.yml:532]: Thanks for checking! Just to verify, the topic is indeed at the intended level in the TOC (as a standalone assembly under "Cluster administration").

File: modules/risks-setting-higher-process-id-limits.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (bhardesty) [reply modules/risks-setting-higher-process-id-limits.adoc:19]: Thanks for the verification - it's certainly an awkward presentation, but there's no good way around the "no xrefs in modules" rule.

File: rosa_cluster_admin/rosa-configuring-pid-limits.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bhardesty) [reply rosa_cluster_admin/rosa-configuring-pid-limits.adoc:26]: In this case, it's just for the preceding module (not the entire assembly). I'll remove the extraneous id.

File: modules/understanding-process-id-limits.adoc (hunk 4 above)
Comment (bhardesty) [reply modules/understanding-process-id-limits.adoc:17]: FYI, re: the link to "node resources" - I changed the module file name to start with the `sd-` prefix. The reason for the link is because it points to a topic that is not included in the ROSA docs. Because we don't have a good way to handle these scenarios, we typically just link to the OCP docs on the customer portal. I renamed the file to indicate that this is a Managed OpenShift (i.e. "service delivery") module, and is not intended for OCP.

**Comment by bhardesty** [comment]
@bburt-rh @shipsing This is ready for you to take another look. Also, I realize it's been several weeks; if it would be better, we can send it through peer review again so someone on the current peer review squad can review.

**Comment by bburt-rh** [comment]
@bhardesty - Can you please send it through for fresh review from the current squad? Thanks!

File: snippets/pid-limits.adoc [hunk 21]
```diff
@@ -15,9 +15,17 @@
 
 :_mod-docs-content-type: SNIPPET
 
+// The PID limits are only configurable through the ROSA CLI. Once they are configurable through OCM, the conditionalization can be removed.
+
 [IMPORTANT]
 ====
-As of the {product-title} versions 4.8.35, 4.9.26, 4.10.6, the {product-title} default per-pod pid limit is `4096`. If you want to enable this PID limit, you must upgrade your {product-title} clusters to these versions or later. {product-title} clusters with prior versions use a default PID limit of `1024`.
+As of {product-title} 4.11, the default per-pod PID limit is `4096`. If you want to enable this PID limit, you must upgrade your {product-title} clusters to this version or later. {product-title} clusters running on earlier versions use a default PID limit of `1024`.
```
Comment (JoeAldinger) [inline snippets/pid-limits.adoc:22]: I'm not sure what the procedure is but I would checkin with px on this support statement.

**Review by JoeAldinger — COMMENTED** [review]
A few suggestion. Overall, good PR and clear prose.

File: modules/risks-setting-higher-process-id-limits.adoc (hunk 7 above)
Comment (bhardesty) [reply modules/risks-setting-higher-process-id-limits.adoc:15]: @robpblake Does this example make sense? What does the 1,048,576 number represent here? How could it be this high if we only support a max of 16,384?

File: snippets/pid-limits.adoc (hunk 21 above)
Comment (bhardesty) [reply snippets/pid-limits.adoc:22]: Thanks - PX confirmed that this should be fine.

File: modules/risks-setting-higher-process-id-limits.adoc (hunk 7 above)
Comment (robpblake) [reply modules/risks-setting-higher-process-id-limits.adoc:15]: @bhardesty I'm fairly sure there is no specific meaning to the `1,048,576` number. I think it was just an example to show how to work out roughly how many pods could run on a single node given the `podPidsLimit` you apply.

I agree that we should use an example with a `podPidsLimit` <= 16,384 as that the upper bound by default.

> How could it be this high if we only support a max of 16,384?

For certain customers, we have the ability to assign a capability to their organization (`capability.organization.bypass_pids_limits`) that allows them to exceed the default 16,384 limit (up to about 3.5. million). This was required as certain important customers already have workloads that go far beyond the 16,384 requirement and asking them to re-architect their workloads to use this feature is not feasible.

So in summary:

- Most customers will stick with the default range and be encouraged to correctly architect their applications to work with this
- For some customers, we'll let them exceed the default limit, particularly if there are legacy workloads that cannot be easily migrated.

cheers,

Rob

Comment (bhardesty) [reply modules/risks-setting-higher-process-id-limits.adoc]: I updated the example to use the max PID limit of 16,384.

**Comment by bscott-rh** [comment]
Merge review LGTM.
